README.md
# CritterForecast

Displays currently available critters.

CHANGELOG.md
# Change Log
All notable changes to the "CritterForecast" extension will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Sortable table.

## [2.1.0]
### Added
- Checklist functionality.

## [2.0.1]
### Fixed
- Hemisphere switching.

## [2.0.0]
### Changed 
- Merge critter routers into single handler.
- Replaced sqlite3 db with flat json db.
### Removed
- sqlite3

## [1.4.1]
### Changed
- Site no longer redirects.
- Minor meta and visual tweaks.
### Fixed
- Service work no longer caches all requests.

## [1.4.0]
### Added
- Google Analytics and Robots.txt
- Service Worker
### Changed
- Web Manifest to be android home screen compliant.
### Removed
- Unncessary enviorment icons.

## [1.3.0]
### Changed
- Rounder Redesign
- Merged Note into Location
- Added locale formating to Price

## [1.2.3]
### Fixed
- Browser no longer caches critter JSON.
- Static files now set to cache.

## [1.2.2]
### Fixed
- Day of Week now properly displays.

## [1.2.1] - 2020-04-17
### Fixed
- Clock now properly updates.

## [1.2.0] - 2020-04-17
### Changed
- Updates no longer reload page.
- New Theme and Images.

## [1.0.0] - 2020-04-16
### Added
- Display currently available critters.
- Highlight last chance opportunities.
- Clock display.
- Hemisphere support.

[Unreleased]: https://github.com/CatStarwind/CritterForecast/compare/v2.1.0...HEAD
[2.1.1]: https://github.com/CatStarwind/CritterForecast/compare/v2.0.1...v2.1.0
[2.0.1]: https://github.com/CatStarwind/CritterForecast/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/CatStarwind/CritterForecast/compare/v1.4.1...v2.0.0
[1.4.1]: https://github.com/CatStarwind/CritterForecast/compare/v1.4.0...v1.4.1
[1.4.0]: https://github.com/CatStarwind/CritterForecast/compare/v1.3.0...v1.4.0
[1.3.0]: https://github.com/CatStarwind/CritterForecast/compare/v1.2.3...v1.3.0
[1.2.3]: https://github.com/CatStarwind/CritterForecast/compare/v1.2.2...v1.2.3
[1.2.2]: https://github.com/CatStarwind/CritterForecast/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/CatStarwind/CritterForecast/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/CatStarwind/CritterForecast/compare/v1.0.0...v1.2.0
[1.0.0]: https://github.com/CatStarwind/CritterForecast/releases/tag/v1.0.0

app.js
var express = require('express');
var path = require('path');
var logger = require('morgan');

var cookieParser = require('cookie-parser');

var favicon = require('serve-favicon');
var rfs = require('rotating-file-stream');

var app = express();
var critterHandler = require('./critters');
// view engine setup
// app.set('views', path.join(__dirname, 'views'));
// app.set('view engine', 'pug');

// Create Rotating Log
var appLogStream = rfs.createStream('critterforecast.log',{
  interval: '1d',
  path: "/data/log"
});
logger.token('remote-addr', function (req, res) { return req.headers['x-real-ip']; });
app.use(logger(':remote-addr [:date[iso]] :status :res[content-length] ":method :url HTTP/:http-version" ":referrer" ":user-agent"', { stream: appLogStream }));

app.use(cookieParser());
app.use(express.json());
app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
app.use(express.urlencoded({ extended: false }));

app.use(express.static(path.join(__dirname, 'public'), {maxAge: '4h'}));
app.use('/fish', critterHandler('fish'));
app.use('/insect', critterHandler('insect'));

module.exports = app;

critters.js
const { DateTime } = require("luxon");
const db = require("./db/critter.json");
const north = (() => { let y = []; for(var i=0;i<12;i++){y.push(2**i);} return y; })();
const south = north.map((n,i) => (i<6) ? n << 6 : n >> 6);
const critters = {
	'insect': {
		'timeframe': [[4,8],[8,16],[16,17],[17,19],[19,22],[22,4]],
		'db': db.insect
	},
	'fish': {
		'timeframe': [[4,9],[9,16],[16,21],[21,4]],
		'db': db.fish
	}
};

module.exports = function(critter) {
	return function(req, res) {
		res.setHeader('Cache-Control', 'no-cache, no-store');
		let db = critters[critter].db;
		let available = [];	

		// Where is Now?
		let month = (req.query.hemi === "South") ? south : north;
		let dt = DateTime.utc();
		let offset = dt.setZone(req.query.tz);
		if(offset.isValid) { dt = offset; }
		
		//Parameters
		let currentMonth = month[dt.month-1];
		let currentTime = 2 ** critters[critter].timeframe.findIndex(tf => (dt.hour >= tf[0] && dt.hour < tf[1]) || (tf[0] > tf[1] && (dt.hour >= tf[0] || dt.hour < tf[1])));
		let monthLC = dt.month < 12 ? month[dt.month] : 1;
		let timeLC = currentTime < 2 ** (critters[critter].timeframe.length-1) ? currentTime * 2 : 1;

		//Create Display Header Index
		let hi = Object.assign({}, db.cols);
		["Month","Time"].forEach(v => delete hi[v]);
		Object.keys(hi).forEach((v,i) => hi[v]=i);

		//Find
		db.rows.forEach((row,i) => {
			if (row[[db.cols.Month]] & currentMonth && row[[db.cols.Time]] & currentTime) {
				let c = [];
				for(let h in hi){ c.push(row[db.cols[h]]); }
				c.push(row[[db.cols.Month]] & monthLC ? 0 : 1);
				c.push(row[[db.cols.Time]] & timeLC ? 0 : 1);
				available.push(c);
			}
		});
		["MLC","TLC"].forEach(v => hi[v] = Object.keys(hi).length);
		
		// Sort Location, Price Desc, ID
		available.sort((a,b) => {
			let locA = a[hi.Location].toUpperCase();
			let locB = b[hi.Location].toUpperCase();

			if(locA < locB) { return -1; }
			if(locA > locB) { return 1; }
			return b[hi.Price] - a[hi.Price] || a[hi.ID] - b[hi.ID];
		});
		
		//Insert Header
		available.unshift(hi);

		res.json(available);
	};
};

public/sw.js
var CACHE_NAME = 'cf-cache-v1';
var urlsToCache = [
	'/',
	'https://fonts.googleapis.com/css2?family=Raleway:wght@800&family=Varela+Round&display=swap',
	'/css/site.css',
	'/js/jquery.min.js',
	'/js/luxon.min.js',
	'/js/site.js',
	'/img/logo.png',
	'/img/critters.png',
	'/img/hemi_north.png',
	'/img/hemi_south.png'
];

self.addEventListener('install', function(event) {
	self.skipWaiting();

	// Perform install steps
	event.waitUntil(
		caches.open(CACHE_NAME).then(function(cache) {
			//console.log('Opened cache');
			return cache.addAll(urlsToCache);
		})
	);
});

self.addEventListener('activate', function(event) {  
	event.waitUntil(
	  caches.keys().then(function(cacheNames) {
		return Promise.all(cacheNames.map(function(cacheName) { return caches.delete(cacheName); }));
	  })
	);
  });

self.addEventListener('fetch', function(event) {
	event.respondWith(
		caches.match(event.request).then(function(response) {
			// Cache hit - return response
			if (response) { return response; }

			return fetch(event.request);
		})
	);
  });

routes/fish.js
const express = require('express');
const router = express.Router();
const sqlite3 = require('sqlite3').verbose();
const { DateTime } = require("luxon");
const db = new sqlite3.cached.Database('./db/critter.db');
process.on('SIGINT', () => { db.close(); });

// Morning, Day, Evening, Night [start,end]
const timeframe = [[4,9],[9,16],[16,21],[21,4]];

router.get('/', function(req, res) {
	res.setHeader('Cache-Control', 'no-cache, no-store');	
	let fish = [];
	let month = (() => { let y = []; for(var i=0;i<12;i++){y.push(2**i);} return y; })();	
	let dt = DateTime.utc();
	let offset = dt.setZone(req.query.tz);
	if(req.query.hemi === "South"){ month.forEach((y,i) => month[i] = (i<6) ? y << 6 : y >> 6); }
	if(offset.isValid) { dt = offset; }

	//Query Parameters
	let currentMonth = month[dt.month-1];
	let currentTime = 2 ** timeframe.findIndex(tf => (dt.hour >= tf[0] && dt.hour < tf[1]) || (tf[0] > tf[1] && (dt.hour >= tf[0] || dt.hour < tf[1])));
	let monthLC = dt.month < 12 ? month[dt.month] : 1;
	let timeLC = currentTime < 2 ** (timeframe.length-1) ? currentTime * 2 : 1;

	db.each(`
		SELECT Name, Price, Location, Note, Size S
			,CASE WHEN Month & ? THEN 0 ELSE 1 END MLC
			,CASE WHEN Time & ? THEN 0 ELSE 1 END TLC
		FROM Fish
		WHERE Month & ? != 0 AND Time & ? != 0
		ORDER BY Location, Price DESC
		`, [monthLC, timeLC, currentMonth, currentTime],
		(err, row) => fish.push(row),
		(err, count) => res.json(fish)
	);
});

module.exports = router;

routes/insect.js
const express = require('express');
const router = express.Router();
const sqlite3 = require('sqlite3').verbose();
const { DateTime } = require("luxon");
const db = new sqlite3.cached.Database('./db/critter.db');
process.on('SIGINT', () => { db.close(); });

// Morning, Day, LateDay, EarlyEvening, Night [start,end]
const timeframe = [[4,8],[8,16],[16,17],[17,19],[19,22],[22,4]];

router.get('/', function(req, res) {
	res.setHeader('Cache-Control', 'no-cache, no-store');	
	let insect = [];
	let month = (() => { let y = []; for(var i=0;i<12;i++){y.push(2**i);} return y; })();	
	let dt = DateTime.utc();
	let offset = dt.setZone(req.query.tz);
	if(req.query.hemi === "South"){ month.forEach((y,i) => month[i] = (i<6) ? y << 6 : y >> 6); }
	if(offset.isValid) { dt = offset; }

	//Query Parameters
	let currentMonth = month[dt.month-1];
	let currentTime = 2 ** timeframe.findIndex(tf => (dt.hour >= tf[0] && dt.hour < tf[1]) || (tf[0] > tf[1] && (dt.hour >= tf[0] || dt.hour < tf[1])));
	let monthLC = dt.month < 12 ? month[dt.month] : 1;
	let timeLC = currentTime < 2 ** (timeframe.length-1) ? currentTime * 2 : 1;

	db.each(`
		SELECT Name, Price, Location, Note
			,CASE WHEN Month & ? THEN 0 ELSE 1 END MLC
			,CASE WHEN Time & ? THEN 0 ELSE 1 END TLC
		FROM Insect
		WHERE Month & ? != 0 AND Time & ? != 0
		ORDER BY Location, Price DESC
		`, [monthLC, timeLC, currentMonth, currentTime],
		(err, row) => insect.push(row),
		(err, count) => res.json(insect)
	);
});

module.exports = router;


public/js/site.js
const timeframe = function(blocks, labels){
	let tf = {};
	
	tf.frameStart = function (hour) {
		let tfstart = [];
		blocks.forEach(b => tfstart.push(b[0]));
		return tfstart.find(fs => fs === hour);
	};

	tf.currentFrame = function (hour) {
		let tfname = "";

		for(var i=0; i<blocks.length; i++){
			let tf = blocks[i];
			if (hour >= tf[0] && hour < tf[1] || (tf[0] > tf[1] && (hour >= tf[0] || hour < tf[1]))) {
				tfname = labels[i];
				break;
			}
		}

		return tfname;
	};

	return tf;
};

const timeframes = {
	"Insect": timeframe([[4,8],[8,16],[16,17],[17,19],[19,22],[22,4]], ["Morning", "Day", "LateDay", "EarlyEvening", "Evening", "Night"]),
	"Fish": timeframe([[4,9],[9,16],[16,21],[21,4]], ["Morning", "Day", "Evening", "Night"])
};

const cf = {
	setCurrentTimeFrame: function(hour) {
		//Clear
		$(".now").removeClass("now");
	
		//Set
		for(let tf in timeframes){
			$("."+timeframes[tf].currentFrame(hour).toLowerCase(), "#"+tf).addClass("now"); 
		}
	},
	updateClock: function(dt) {
		if(dt.toFormat("h:mm a") !== $(".time").text()){
			$(".time").html(dt.toFormat("h:mm") + `<span> ${dt.toFormat("a")}</span>`);
			$(".date").html(dt.toFormat("LLLL dd") + ` <span>${dt.toFormat("ccc")}</span>`);
			
			//Top of Hour Updates
			if (dt.minute === 0) {
				//Clear & Set
				$("#acClock").removeClass().addClass(timeframes.Insect.currentFrame(dt.hour).toLowerCase());

				//Check if entering new Time Frame
				for(let tf in timeframes) {
					if (timeframes[tf].frameStart(dt.hour)) {
						this.createTable(tf);
						this.setCurrentTimeFrame(dt.hour);
					}
				}
			}
		}	
	},
	createTable: function (critter) {
		let id = "#" + critter;
		let url = critter.toLowerCase() + "?tz=" + localStorage.getItem("tz") + "&hemi=" + localStorage.getItem("hemi");
		
		$.getJSON(url, function(data){
			$("tbody", id).empty(); //Reset
			$("th.catch", id).remove();
			let col = data.shift();
			
			data.forEach(row => {
				let $row = $("<tr>").attr("id", critter[0] + row[col.ID]);
				if (row[col.TLC]) { $row.addClass("lastchanceTime"); }
				if (row[col.MLC]) { $row.addClass("lastchanceMonth"); }

				$("<td>").text(row[col.Name]).appendTo($row);
				$("<td>").text(row[col.Price].toLocaleString()).appendTo($row);
				$("<td>").text(row[col.Location] + (row[col.Note] ? ` (${row[col.Note]})` : '')).appendTo($row);
				if(row[col.S]) { $("<td>").text(row[col.S]).appendTo($row); }
				
				$("table", id).append($row);
			});

			if (cf.caught.disp()) { cf.caught.show(id); }
		});
	},
	setHemi: function(hemi) {
		$("#hemisphere").attr({
			src: `/img/hemi_${hemi.toLowerCase()}.png`,
			alt: hemi + " Hemisphere",
			title: hemi + " Hemisphere"
		});
	},
	switchHemi: function() {
		let hemi = localStorage.getItem("hemi") === "North" ? "South" : "North";
		localStorage.setItem("hemi", hemi);
		this.setHemi(hemi);
		for(var tf in timeframes) { cf.createTable(tf); }
	},
	caught: {
		disp: () => localStorage.getItem("catchDisp") === "false" ? false : true,
		list: function() {
			let caught = localStorage.getItem("caught");
			if(!caught) { return []; }
			return caught.split(",");
		},
		check: id => cf.caught.list().includes(id),
		show: function(table) {
			$(table).addClass("showCatch").removeClass("hideCatch");
			$("thead tr", table).prepend('<th class="catch">');
			$("tbody tr", table).each((i,el) => {
				let id = $(el).attr("id");
				if(this.check(id)) { $(el).addClass("caught"); }
				$("<td>").addClass("catch").click(cf.caught.critter).prependTo(el);
			});
		},
		toggle: function() {
			if(!cf.caught.disp()){
				$("div.critter table").each((i,el) => { cf.caught.show(el); });
				$("#catchToggle").removeClass("hide");
				localStorage.setItem("catchDisp", "true");
			} else {
				$(".catch").remove();
				$("div.critter table").addClass("hideCatch").removeClass("showCatch");
				$("#catchToggle").addClass("hide");
				$(".caught").removeClass("caught");
				localStorage.setItem("catchDisp", "false");
			}
		},
		critter: function(e) {
			let id = $(this).parent("tr").attr("id");
			let caught = cf.caught.list();

			if(cf.caught.check(id)){
				caught.splice(caught.indexOf(id), 1)
				localStorage.setItem("caught", caught);
				$(this).parent("tr").removeClass("caught");
			} else {
				caught.push(id);
				localStorage.setItem("caught", caught);
				$(this).parent("tr").addClass("caught");
			}
		}
	},
	init: function() {
		let dt = luxon.DateTime.local();
		this.updateClock(dt);
		this.setCurrentTimeFrame(dt.hour);
		this.setHemi(localStorage.getItem("hemi"));
		for(var tf in timeframes) { this.createTable(tf); }

		$("#catchToggle").toggleClass("hide", !this.caught.disp());
		$("#acClock").addClass(timeframes.Insect.currentFrame(dt.hour).toLowerCase()).attr("title", localStorage.getItem("tz"));
		$("#hemisphere").click(() => cf.switchHemi());
		$("#catchToggle").click(() => cf.caught.toggle());
	}
};

$(function() {
	cf.init();
	
	cf.tick = window.setInterval(function () {
		cf.updateClock(luxon.DateTime.local());
	}, 1000);
});